### PR TITLE
Add page enumerator and access to the rate limit in the cursor

### PR DESCRIPTION
**Pull request by mzaccari**
This builds on @lfittl's pull request, https://github.com/sferik/twitter/pull/626, and uses sferik's comments to hopefully improve the interfaces between the `Enumerable`, `Cursor`, and `RateLimit` classes. With this we can check the state of the rate limit to ensure we don't exceed it:

```ruby
client.follower_ids(count: 500).each_page_with_cursor do |follower_ids, cursor|
  # do something with follower_ids

  break if cursor.rate_limit.reached?
end
```

I tried to DRY up the rpsec tests as best I could for the `Enumerable` classes. Any feedback is appreciated.

### CODE REVIEW

**Comment by sferik** [comment]
Overall, this looks good.

My only concern is that the new `Cursor#rate_limit` method is not threadsafe. Any thoughts about that?

**Comment by mzaccari** [comment]
How does this look: https://github.com/mzaccari/twitter/commit/1edd003af6c8a364f18b115f0d32f46ec238fc97

**Comment by mzaccari** [comment]
The more I use this, the more I think one of 2 possible routes might be better:

1) Expose the `request` object instead of the `rate_limit`, as it is more generic and then cursors don't have to know about rate limits

or

2) Track rate limiting as a separate, memoizable object on the client, incrementing/decrementing what we know of (through requests during the current session) or just fetching the rate limit data from twitter

With 1, we can expose the `request` object for single requests like for a user's profile, which is a single call but still uses a rate limit. We would only have to say something like:

```ruby
client = user.twitter_client
user, request = client.user_with_request
user.screen_name # => 'michaelzaccari'
request.rate_limit.remaining # => 14
```

The cursor, search results, and other enumerables should also have request access:

```ruby
client = user.twitter_client
client.search('#NASA').each_page_with_request |tweets, request|
  tweets.each do |tweet|
    # Do something with tweet
  end
  break if request.rate_limit.reached?
end
```

With 2, a cache of the latest rate limit data could be stored, with fetches made to the API:

```ruby
client = user.twitter_client
rate_limit = client.rate_limit # => { 'search' => {"limit": 15, "remaining": 15, "reset": 1403602426 } }
```

Or something similar. 

If you have the time I'd like to see if these routes might be better before merging this in.

**Comment by mzaccari** [comment]
The cursor object does not change in between pages. This can be easily tested by printing `self.object_id` in the `fetch_next_page` method in the `Cursor` class for multiple requests. You can then run:

```ruby
client.follower_ids(count: 3).each_with_index do |id, i|
  break if i > 12
end
```

Also, I've added a synchronize around setting the request

**Comment by sferik** [comment]
@mzaccari Here is my concern around synchronization:

```ruby
client = Twitter::REST::Client.new(credentials)
cursor1 = client.followers
cursor2 = client.followers
cursor1.take(100)
cursor1.rate_limit.remaining # 10
cursor2.take(100)
cursor2.rate_limit.remaining # 5
cursor1.rate_limit.remaining # 10 <- this value is wrong
```

I could imagine race conditions in multiple threads where this becomes an issue but, as you can see, it can even prove problematic in a single thread.

I also have some concerns about the growing inconsistencies between `Twitter::Cursor` and the other classes that mixin `Twitter::Enumerable` (namely, `Twitter::SearchResults`, `Twitter::TrendResults`, and `Twitter::GeoResults`). What’s the argument for adding these methods only to `Twitter::Cursor`? It seems like they would be useful when paging through search results as well, no?

File: lib/twitter/enumerable.rb [hunk 1]
```diff
@@ -2,16 +2,25 @@ module Twitter
   module Enumerable
     include ::Enumerable
 
+    METHODS = [:each, :each_page].freeze
```
Comment (sferik) [inline lib/twitter/enumerable.rb]: What purpose does this constant serve? I don’t see it used anywhere in the code. It seems like a case of [test-induced design damage](http://david.heinemeierhansson.com/2014/test-induced-design-damage.html). If it’s only used in the tests, this constant belongs in the tests, not in the code.

**Comment by mzaccari** [comment]
> I could imagine race conditions in multiple threads where this becomes an issue but, as you can see, it can even prove problematic in a single thread.

I agree, which is why I am starting to lean towards a solution that encapsulates the `RateLimit` with a singleton that is updated after each request (via the `Request` object) and when a stale entry is missing we can fetch the user's latest rate limit [from here](https://dev.twitter.com/rest/reference/get/application/rate_limit_status).

I am willing to decline this pull request in pursuit of a more `RateLimit`-centric solution. Let me know what you think

**Pull request closed without merging**